File: Assets/Scripts/Inventory.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class InventoryItem{
    public string prefabName;
    public int amount;

    public InventoryItem(string prefabName, int amount){
        this.prefabName = prefabName;
        this.amount = amount;
    }
}


public class Inventory{
    public List<InventoryItem> inventoryList = new List<InventoryItem>();
    public Inventory(){

    }

    public bool itemPresent(string name){
        foreach (InventoryItem item in inventoryList)
        {
            if(item.prefabName == name){
                return true;
            }
        }
        return false;
    }
    public InventoryItem getItem(string name){
        foreach (InventoryItem item in inventoryList)
        {
            if(item.prefabName == name){
                return item;
            }
        }
        return null;
    }
    public void addItem(string name, int amount=1){
        if(itemPresent(name)){
            getItem(name).amount+=amount;
        }else{
            InventoryItem newItem = new InventoryItem(name, amount);
            inventoryList.Add(newItem);
        }
    }
    public void removeItem(string name, int amount=1){
        InventoryItem item = getItem(name);
        item.amount-=amount;
        if(item.amount<=0){
            inventoryList.Remove(item);
        }
    }
}

File: Server/Utils/Inventory.js
const InventoryItem = require("./InventoryItem");
class Inventory {
    constructor() {
        this.inventoryList = [];
    }

    itemPresent(name) {
        //console.log(this.getItem(name))
        if (this.getItem(name) != null) {
            return true;
        } else {
            return false;
        }
    }
    getItem(name) {
        var item = null;
        this.inventoryList.forEach(inventoryItem => {
            if (inventoryItem.prefabName == name) {
                item = inventoryItem;
            }
        });
        return item
    }
    addItem(name, amount, type) {
        if (this.itemPresent(name)) {
            this.getItem(name).amount += amount;
        } else {
            var newItem = new InventoryItem(name, amount, type);
            this.inventoryList.push(newItem);
        }
    }
    removeItem(name, amount = 1) {
        var item = this.getItem(name)
        if (item) {
            item.amount -= amount;
            if (item.amount <= 0) {
                this.inventoryList.splice(this.inventoryList.indexOf(item), 1);
            }
        } else {
            console.warn("No such item in inventory");
        }

    }
    Serialize() {
        return JSON.stringify(this);
    }
    static Deserialize(jsonString) {
        var newInv = new Inventory();
        newInv.inventoryList = JSON.parse(jsonString).inventoryList;
        return newInv;
    }
}

module.exports = Inventory;